File: find_kth_smallest_element_given_ranges.cpp
//{ Driver Code Starts
/* Driver program to test above function */

#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++

class Solution{
    public:
    vector<int>kthSmallestNum(vector<vector<int>>&range, vector<int>queries){
        //Write your code here
        sort(range.begin(), range.end());
        vector<vector<int>> r;
        for(int i = 0; i < range.size(); ++i) {
            if (r.empty()) {
                r.push_back(range[i]);
                continue;
            }
            auto current = range[i];
            while(!r.empty()) {
                auto c = r[r.size()-1];
                if (c[0] >= current[0] && c[0] <= current[1]
                    || c[1] >= current[0] && c[1] <= current[1]
                    || current[0] >= c[0] && current[0] <= c[1]
                    || current[1] >= c[1] && current[1] <= c[1]) {
                        current[0] = min(c[0], current[0]);
                        current[1] = max(c[1], current[1]);
                        r.pop_back();
                } else {
                    break;
                }
            }
            
            r.push_back(current);
        }
        vector<int> ans;
        for(int q : queries) {
            for(int j = 0; j < r.size(); ++j) {
                if (q <= r[j][1]-r[j][0]+1) {
                    ans.push_back(r[j][0] + q - 1);
                    q = -1;
                    break;
                } else {
                    q -= r[j][1]-r[j][0]+1;
                }
            }
            if (q != 0 && q!=-1) {
                ans.push_back(-1);
            }
        }
        return ans;
    } 
};


//{ Driver Code Starts.
int main()
{
	int t;
	cin>>t;
	while(t--)
	{
	    int n;
	    cin >> n;
	    vector<vector<int>>range(n, vector<int>(2, 0));
	    for(int i = 0 ; i < n; i++){
	        cin >> range[i][0] >> range[i][1];
	    }
	    int q;
	    cin >> q;
	    vector<int>queries;
	    for(int i = 0 ; i < q; i++){
	        int x;
            cin >> x;
	        queries.push_back(x);
	    }
	    Solution ob;
	    vector<int>ans = ob.kthSmallestNum(range, queries);
	    for(auto it : ans){
	        cout << it << " ";
	    }
	    cout << endl;
	}
	return 0;
}

// } Driver Code Ends

File: misc/find_kth_smallest_element_given_ranges.py
from typing import List


class Solution:
    def interval(self, r1, r2):
        if r1[0] > r2[1] or r1[1] < r2[0]:
            return False
        return True

    def kthSmallestNum(
        self, n: int, ranges: List[List[int]], q: int, queries: List[int]
    ) -> List[int]:
        # code here
        ranges.sort()
        nr = []
        for r in ranges:
            temp = r
            while len(nr) > 0 and self.interval(temp, nr[-1]):
                temp = [min(nr[-1][0], r[0]), max(nr[-1][1], r[1])]
                nr.pop()
            nr.append(temp)

        ans = []
        for query in queries:
            for r in nr:
                if query <= r[1] - r[0] + 1:
                    ans.append(r[0] + query - 1)
                    query = 0
                    break
                else:
                    query -= r[1] - r[0] + 1

            if query != 0:
                ans.append(-1)

        return ans


# {
# Driver Code Starts


class IntMatrix:
    def __init__(self) -> None:
        pass

    def Input(self, n, m):
        matrix = []
        # matrix input
        for _ in range(n):
            matrix.append([int(i) for i in input().strip().split()])
        return matrix

    def Print(self, arr):
        for i in arr:
            for j in i:
                print(j, end=" ")
            print()


class IntArray:
    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  # array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        n = int(input())

        ranges = IntMatrix().Input(n, 2)

        q = int(input())

        queries = IntArray().Input(q)

        obj = Solution()
        res = obj.kthSmallestNum(n, ranges, q, queries)

        IntArray().Print(res)


# } Driver Code Ends
